### ISSUE TITLE
need throwsCastError

### ISSUE THREAD

**Issue by trinarytree** [dart-lang/test#904]
throws_matchers.dart seems like it ought to include matchers for every standard throwable, including `CastError`.  currently it doesn't have such a matcher.

a few other standard errors that also don't have a corresponding matcher:
```
AbstractClassInstantiationError
AssertionError
AsyncError
FallThroughError
JsonUnsupportedObjectError
OutOfMemoryError
RemoteError
StackOverflowError
```
it's not clear why the 9 matchers currently present in throws_matchers.dart are there and not others.  maybe someone wanted to wait until someone complained before adding new ones?

**Comment by natebosch** [comment]
I think for the most part we tended to have matchers for the errors that were more commonly things you'd want to test for - things that represent bugs in _usage_ (so we'd expect the API contract to throw on them) rather than bugs in _implementation_ (so we'd expect the code to be fixed).

I might design an API which is intended to throw `ArgumentError` for some incorrect usages, but I'd never _intend_ to throw `OutOfMemoryError` or `StackOverflowError`. I'm a little unsure about `CastError`, I would not expect to design an API that expects to be able to make a certain cast without encoding that in the signature, or checking myself and throwing an `ArgumetnError` instead... On the other hand with Dart 2 I wouldn't be shocked if someone was designing such an API contract.

Can you explain your use case for wanting to test that a given call throws a `CastError`? Is it painful for you to stick the implementation of `isCastError` next to the test?

**Comment by trinarytree** [comment]
i see a bunch of different projects that each roll their own `throwsCastError`, e.g.
- the expect library
- the async library
- collection/test/utils.dart
- my project now has a use case where we have a piece of infra that transforms types in a contractual way, mostly to overcome Dart 2 finickiness.  a test for that infra ensures that it correctly throws a `CastError` if client code uses it incorrectly.

at a higher level, Dart differentiates "exceptions" and "errors", the latter representing programmer error. at first blush, one might think that there is no reason for a test to guarantee that an error is thrown, but infra code may promise to catch certain kinds of programmer errors and so its authors would want to test that.

**Comment by natebosch** [comment]
Sounds good. Let's add `throwsCastError` and we'll hold off on the others until a concrete use case comes up.

**Comment by vargavince91** [comment]
Would you consider a generic solution (inspired https://github.com/dart-lang/sdk/issues/39305#issuecomment-580732718):

```dart
Matcher throwsTypeOf<T>() => throwsA(TypeMatcher<T>());
```

**Comment by natebosch** [comment]
We discussed this offline. I don't think we want a special case for this, we have too many special cases already.

tl;dr;

In the long term the usage we hope for is `throws(isA<CastError>())`, and we hope to _always_ use that pattern.

In the short term I think we want to go with `throwsA(isA<CastError>())`.

In the next breaking release we'll drop `throws` altogether. It has been deprecated for a while. We'll wait for confidence that code has migrated, then we'll reintroduce it it as `Matcher throws([Matcher matcher = anything]) => throwsA(matcher);`. The way get the behavior of the old `throws` will be `throws()`. The type on the argument should alleviate the confusion surfaced in https://github.com/dart-lang/sdk/issues/39305 since a `Type` is not a `Matcher`. The equivalent to `throwsA(someValue)` will be `throws(equals(someValue))`. 

In the very long term, if we can handle the churn, we'd like to deprecated and remove all the `is*` and `throws*` variables, in favor of _always_ using `throws()` and `isA<SomeType>()` for those use cases.
